hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java

```diff
@@ -522,7 +528,9 @@ public Void visitFromClause(SqmFromClause sqmFromClause) {
 	protected void consumeFromClauseRoot(SqmRoot<?> sqmRoot) {
 		log.tracef( "Resolving SqmRoot [%s] to TableGroup", sqmRoot );
 
-		assert ! fromClauseIndex.isResolved( sqmRoot );
+		if ( fromClauseIndex.isResolved( sqmRoot ) ) {
+			log.tracef( "Already resolved SqmRoot [%s] to TableGroup", sqmRoot );
+		}
```

Comment:
@sebersole I had to suppress this assertion to get `select` clause subqueries referenced in the `order by` clause to work. But I suppose there must be some deeper cause to this that is perhaps worth fixing.

Reply:
No idea why that would be necessary.  It would seem like you are resolving the from-clause twice or perhaps related to "scoping"

Reply:
@sebersole Yeah it confused me too, so I went digging the other day and I believe it's related to the fact that by default a select expression referenced by alias in the `order by` clause gets copied into the `order by`.  Now this behavior surprised me, because I had been under the impression that most databases allow you to reference a selected expression by alias in `order by`, but somebody wrote explicit code to do this so there must be a reason for it.

Anyway it means that `select` clause expression can end up being walked twice.

Reply:
> but somebody wrote explicit code to do this so there must be a reason for it

FTR, it looks like "somebody" was you haha.

Reply:
You mean something like HQL `select p from Person p order by p`?

But the way this is meant to work is that `#consumeFromClauseRoot` (this method) only happens from `#visitFromClause`.  All other cases route through `#visitRootPath`.

Basically, when walking the from-clause nodes we do a specialized "visitation".  Other usages of a `SqmFrom` in other parts of the query-spec are handled by the normal visitation.

Do you have a stack trace?

Reply:
oh, i see what you mean...  For curiosity, if you can still produce that, could you switch `org.hibernate.dialect.Dialect#replaceResultVariableInOrderByClauseWithPosition` to true and see what happens?

Reply:
I mean like this HQL:

    select (select 1) as one, (select 'foo') as foo order by one, foo

Currently produces this SQL:

    select
        (select
            1 
        from
            dual),
        (select
            'foo' 
        from
            dual) 
    from
        dual 
    order by
        (select
            1 
        from
            dual),
        (select
            'foo' 
        from
            dual)

Reply:
@sebersole if I'm not mistaken, `replaceResultVariableInOrderByClauseWithPosition()` appears to currently have no effect.

Reply:
It should affect how `AbstractSqlAstWalker#visitSqlSelectionExpression` works:

```
	@Override
	public void visitSqlSelectionExpression(SqlSelectionExpression expression) {
		final boolean useSelectionPosition = dialect.replaceResultVariableInOrderByClauseWithPosition();

		if ( useSelectionPosition ) {
			appendSql( Integer.toString( expression.getSelection().getJdbcResultSetIndex() ) );
		}
		else {
			expression.getExpression().accept( this );
		}
	}
```

Reply:
I just saw this discussion and wanted to mention that some DBMS i.e. PostgreSQL and I also think MySQL also allow to use a select alias in the GROUP BY clause. Support for that was merged into 5.4 recently.

Reply:
@beikov Yeah, I think that's ideal.

Reply:
@sebersole that method never gets called in my test.